Failed to build afopencl x64 target on MSVS 2017

Hi,

Building `afopencl` target on x64 MSVS 2017 is failing at master d2df83d (5 June 2020) with a bunch of errors relating to "declaration is not allowed in 'constexpr' function body". The culprit appears to be `getTypeBuildDefinition()` defined in `\arrayfire\src\backend\opencl\types.hpp:128`.

Full error message...

```
10>git\arrayfire\src\backend\opencl\types.hpp(137): error C3546: '...': there are no parameter packs available to expand
10>E:\Russ\SwDev\lu_dev\git\arrayfire\src\backend\opencl\kernel/reduce.hpp(139): note: see reference to function template instantiation 'const char *opencl::getTypeBuildDefinition<Ti,To>(void)' being compiled
10>        with
10>        [
10>            Ti=float,
10>            To=char
10>        ]
10>E:\Russ\SwDev\lu_dev\git\arrayfire\src\backend\opencl\kernel/reduce.hpp(232): note: see reference to function template instantiation 'void opencl::kernel::reduceFirstLauncher<Ti,To,af_and_t>(opencl::Param,opencl::Param,const opencl::uint,const opencl::uint,const opencl::uint,int,double)' being compiled
10>        with
10>        [
10>            Ti=float,
10>            To=char
10>        ]
10>git\arrayfire\src\backend\opencl\types.hpp(135): error C3250: 'is_half': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(136): error C3250: 'is_double': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(139): error C3250: 'half_def': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(141): error C3250: 'double_def': declaration is not allowed in 'constexpr' function body
10>git\arrayfire\src\backend\opencl\types.hpp(128): error C3259: 'constexpr' functions can only have one return statement
```

Russ

Hi,

Sorry I'm new to GitHub pull requests - how do I download a diff for the 5 changed files? I can browse the changed files in commit 62f6ced and I can download the entire branch at commit 62f6ced as a ZIP file, but I don't see a away to download a diff.

When I download the ZIP file and run the following, `cmake` is failing.

```
cmake -T v140 -G "Visual Studio 15 2017 Win64" -DCMAKE_TOOLCHAIN_FILE="git\vpkg\scripts\buildsystems\vcpkg.cmake" ..
```

```
CMake Error: File /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/forge/CMakeModules/version.h.in does not exist.
CMake Error at CMakeModules/AFconfigure_forge_submodule.cmake:47 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  CMakeLists.txt:117 (include)


CMake Error at CMakeLists.txt:163 (add_subdirectory):
  The source directory

    /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/spdlog

  does not contain a CMakeLists.txt file.


CMake Error at CMakeLists.txt:164 (add_subdirectory):
  The source directory

    /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/glad

  does not contain a CMakeLists.txt file.


-- CUDA_architecture_build_targets: 7.0 6.1
-- Found OpenCL: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.2/lib/x64/OpenCL.lib (found version "1.2")
CMake Error at test/CMakeLists.txt:27 (add_subdirectory):
  The source directory

    /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/test/gtest

  does not contain a CMakeLists.txt file.


CMake Error at test/CMakeLists.txt:28 (set_target_properties):
  set_target_properties Can not find target to add properties to: gtest


-- Performing Test pthread_flag
-- Performing Test pthread_flag - Failed
fatal: not a git repository (or any of the parent directories): .git
-- No git. Setting hash to default
-- Configuring incomplete, errors occurred!
See also "/git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/build/CMakeFiles/CMakeOutput.log".
See also "/git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/build/CMakeFiles/CMakeError.log".
```

I assume that I am just trying to use the pull request incorrectly, so any tips would be appreciated.

Russ

That is why I shared the patch URL as hyperlink in my earlier comments.
Just in case, it is posted below again

https://github.com/arrayfire/arrayfire/pull/2925.patch

1.Open the link in your browser.
2. Save the file to disk.
3. now run the command "git apply patch_file-name"

Now run the build command as usual.

On Mon, 8 Jun, 2020, 14:26 Russ, <notifications@github.com> wrote:

> Hi,
>
> Sorry I'm new to GitHub pull requests - how do I download a diff for the 5
> changed files? I can browse the changed files in commit 62f6ced
> <https://github.com/arrayfire/arrayfire/commit/62f6ced8da86b1232c661f0c53765494e1837bbc>
> and I can download the entire branch at commit 62f6ced
> <https://github.com/arrayfire/arrayfire/commit/62f6ced8da86b1232c661f0c53765494e1837bbc>
> as a ZIP file, but I don't see a away to download a diff.
>
> When I download the ZIP file and run the following, cmake is failing.
>
> cmake -T v140 -G "Visual Studio 15 2017 Win64" -DCMAKE_TOOLCHAIN_FILE="git\vpkg\scripts\buildsystems\vcpkg.cmake" ..
>
> CMake Error: File /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/forge/CMakeModules/version.h.in does not exist.
> CMake Error at CMakeModules/AFconfigure_forge_submodule.cmake:47 (configure_file):
>   configure_file Problem configuring file
> Call Stack (most recent call first):
>   CMakeLists.txt:117 (include)
>
>
> CMake Error at CMakeLists.txt:163 (add_subdirectory):
>   The source directory
>
>     /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/spdlog
>
>   does not contain a CMakeLists.txt file.
>
>
> CMake Error at CMakeLists.txt:164 (add_subdirectory):
>   The source directory
>
>     /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/extern/glad
>
>   does not contain a CMakeLists.txt file.
>
>
> -- CUDA_architecture_build_targets: 7.0 6.1
> -- Found OpenCL: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.2/lib/x64/OpenCL.lib (found version "1.2")
> CMake Error at test/CMakeLists.txt:27 (add_subdirectory):
>   The source directory
>
>     /git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/test/gtest
>
>   does not contain a CMakeLists.txt file.
>
>
> CMake Error at test/CMakeLists.txt:28 (set_target_properties):
>   set_target_properties Can not find target to add properties to: gtest
>
>
> -- Performing Test pthread_flag
> -- Performing Test pthread_flag - Failed
> fatal: not a git repository (or any of the parent directories): .git
> -- No git. Setting hash to default
> -- Configuring incomplete, errors occurred!
> See also "/git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/build/CMakeFiles/CMakeOutput.log".
> See also "/git/arrayfire-62f6ced8da86b1232c661f0c53765494e1837bbc/build/CMakeFiles/CMakeError.log".
>
> I assume that I am just trying to use the pull request incorrectly, so any
> tips would be appreciated.
>
> Russ
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/arrayfire/arrayfire/issues/2924#issuecomment-640467451>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAY6OOUSRKPHQ7CQEE5EQGLRVSRTTANCNFSM4NTPNDLA>
> .
>

--- a/CMakeModules/InternalUtils.cmake
+++ b/CMakeModules/InternalUtils.cmake
@@ -172,18 +172,26 @@ macro(arrayfire_set_cmake_default_variables)
   # This code is used to generate the compilers.h file in CMakeModules. Not all
   # features of this modules are supported in the versions of CMake we wish to
   # support so we are directly including the files here
-  # include(WriteCompilerDetectionHeader)
-  # write_compiler_detection_header(
-  #         FILE ${ArrayFire_BINARY_DIR}/include/af/compilers.h
-  #         PREFIX AF
-  #         COMPILERS AppleClang Clang GNU Intel MSVC
-  #         # NOTE: cxx_attribute_deprecated does not work well with C
-  #         FEATURES cxx_rvalue_references cxx_noexcept cxx_variadic_templates cxx_alignas cxx_static_assert cxx_generalized_initializers
-  #         ALLOW_UNKNOWN_COMPILERS
-  #         #[VERSION <version>]
-  #         #[PROLOG <prolog>]
-  #         #[EPILOG <epilog>]
-  #         )
+  #  set(compiler_header_epilogue [=[
+  #  #if defined(AF_COMPILER_CXX_RELAXED_CONSTEXPR) && AF_COMPILER_CXX_RELAXED_CONSTEXPR
+  #  #define AF_CONSTEXPR constexpr
+  #  #else
+  #  #define AF_CONSTEXPR
+  #  #endif
+  #  ]=])
+  #  include(WriteCompilerDetectionHeader)
+  #  write_compiler_detection_header(
+  #          FILE ${ArrayFire_BINARY_DIR}/include/af/compilers.h
+  #          PREFIX AF
+  #          COMPILERS AppleClang Clang GNU Intel MSVC
+  #          # NOTE: cxx_attribute_deprecated does not work well with C
+  #          FEATURES cxx_rvalue_references cxx_noexcept cxx_variadic_templates cxx_alignas
+  #          cxx_static_assert cxx_generalized_initializers cxx_relaxed_constexpr
+  #          ALLOW_UNKNOWN_COMPILERS
+  #          #[VERSION <version>]
+  #          #[PROLOG <prolog>]
+  #          EPILOG ${compiler_header_epilogue}
+  #          )
   configure_file(
     ${CMAKE_MODULE_PATH}/compilers.h
     ${ArrayFire_BINARY_DIR}/include/af/compilers.h)

--- a/CMakeModules/InternalUtils.cmake
+++ b/CMakeModules/InternalUtils.cmake
@@ -29,13 +29,25 @@ endif()
 endfunction()
 
 function(arrayfire_get_cuda_cxx_flags cuda_flags)
-  if(NOT MSVC)
-    set(flags -std=c++14 --expt-relaxed-constexpr -Xcompiler -fPIC -Xcompiler ${CMAKE_CXX_COMPILE_OPTIONS_VISIBILITY}hidden)
-  else()
-    set(flags -Xcompiler /wd4251 -Xcompiler /wd4068 -Xcompiler /wd4275 -Xcompiler /bigobj -Xcompiler /EHsc)
+  if(MSVC)
+    set(flags -Xcompiler /wd4251
+              -Xcompiler /wd4068
+              -Xcompiler /wd4275
+              -Xcompiler /bigobj
+              -Xcompiler /EHsc
+              --expt-relaxed-constexpr)
     if(CMAKE_GENERATOR MATCHES "Ninja")
       set(flags ${flags} -Xcompiler /FS)
     endif()
+    if(cplusplus_define)
+      list(APPEND flags -Xcompiler /Zc:__cplusplus
+                        -Xcompiler /std:c++14)
+    endif()
+  else()
+    set(flags -std=c++14
+              -Xcompiler -fPIC
+              -Xcompiler ${CMAKE_CXX_COMPILE_OPTIONS_VISIBILITY}hidden
+              --expt-relaxed-constexpr)
   endif()
 
   if("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU" AND

--- a/CMakeModules/compilers.h
+++ b/CMakeModules/compilers.h
@@ -202,6 +202,13 @@
 #      define AF_COMPILER_CXX_GENERALIZED_INITIALIZERS 0
 #    endif
 
+#if ((__clang_major__ * 100) + __clang_minor__) >= 400 &&                      \
+    __has_feature(cxx_relaxed_constexpr)
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 1
+#else
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 0
+#endif
+
 #  elif AF_COMPILER_IS_Clang
 
 #    if !(((__clang_major__ * 100) + __clang_minor__) >= 301)
@@ -252,6 +259,13 @@
 #    else
 #      define AF_COMPILER_CXX_GENERALIZED_INITIALIZERS 0
 #    endif
+
+#if ((__clang_major__ * 100) + __clang_minor__) >= 301 &&                      \
+    __has_feature(cxx_relaxed_constexpr)
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 1
+#else
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 0
+#endif
 
 #  elif AF_COMPILER_IS_GNU
 
@@ -306,6 +320,12 @@
 #    else
 #      define AF_COMPILER_CXX_GENERALIZED_INITIALIZERS 0
 #    endif
+
+#if (__GNUC__ * 100 + __GNUC_MINOR__) >= 500 && __cplusplus >= 201402L
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 1
+#else
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 0
+#endif
 
 #  elif AF_COMPILER_IS_Intel
 
@@ -378,6 +398,20 @@
 #      define AF_COMPILER_CXX_GENERALIZED_INITIALIZERS 0
 #    endif
 
+#if __cpp_constexpr >= 201304 ||                                               \
+    (__INTEL_COMPILER >= 1700 &&                                               \
+     ((__cplusplus >= 201300L) ||                                              \
+      ((__cplusplus == 201103L) && !defined(__INTEL_CXX11_MODE__)) ||          \
+      ((((__INTEL_COMPILER == 1500) && (__INTEL_COMPILER_UPDATE == 1))) &&     \
+       defined(__GXX_EXPERIMENTAL_CXX0X__) &&                                  \
+       !defined(__INTEL_CXX11_MODE__)) ||                                      \
+      (defined(__INTEL_CXX11_MODE__) && defined(__cpp_aggregate_nsdmi))) &&    \
+     !defined(_MSC_VER))
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 1
+#else
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 0
+#endif
+
 #  elif AF_COMPILER_IS_MSVC
 
 #    if !(_MSC_VER >= 1600)
@@ -436,6 +470,12 @@
 #      define AF_COMPILER_CXX_GENERALIZED_INITIALIZERS 0
 #    endif
 
+#if _MSC_VER >= 1911
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 1
+#else
+#define AF_COMPILER_CXX_RELAXED_CONSTEXPR 0
+#endif
+
 #  endif
 
 #  if defined(AF_COMPILER_CXX_NOEXCEPT) && AF_COMPILER_CXX_NOEXCEPT
@@ -469,6 +509,13 @@ template<> struct AFStaticAssert<true>{};
 #    define AF_STATIC_ASSERT_MSG(X, MSG) enum { AF_STATIC_ASSERT_JOIN(AFStaticAssertEnum, __LINE__) = sizeof(AFStaticAssert<X>) }
 #  endif
 
+#endif
+
+#if defined(AF_COMPILER_CXX_RELAXED_CONSTEXPR) &&                              \
+    AF_COMPILER_CXX_RELAXED_CONSTEXPR
+#define AF_CONSTEXPR constexpr
+#else
+#define AF_CONSTEXPR
 #endif
 
 #endif

--- a/src/backend/common/half.hpp
+++ b/src/backend/common/half.hpp
@@ -10,12 +10,36 @@
 #pragma once
 
 #if defined(NVCC) || defined(__CUDACC_RTC__)
+
+// MSVC sets __cplusplus to 199711L for all versions unless you specify
+// the new \Zc:__cplusplus flag in Visual Studio 2017. This is not possible
+// in older versions of MSVC so we updated it here for the cuda_fp16 header
+// because otherwise it does not define the default constructor for __half
+// as default and that prevents the __half struct to be used in a constexpr
+// expression
+#if defined(_MSC_VER) && __cplusplus == 199711L
+#undef __cplusplus
+#define __cplusplus 201402L
+#define AF_CPLUSPLUS_CHANGED
+#endif
+
 #include <cuda_fp16.h>
+
+#ifdef AF_CPLUSPLUS_CHANGED
+#undef __cplusplus
+#undef AF_CPLUSPLUS_CHANGED
+#define __cplusplus 199711L
+#endif
 #endif
 
 #include <backend.hpp>
 
-#ifndef __CUDACC_RTC__
+#ifdef __CUDACC_RTC__
+using uint16_t = unsigned short;
+// we do not include the af/compilers header in nvrtc compilations so
+// we are defining the AF_CONSTEXPR expression here
+#define AF_CONSTEXPR constexpr
+#else
 #include <af/compilers.h>
 #include <cstring>
 #include <ostream>
@@ -23,8 +47,6 @@
 #include <type_traits>
 
 #include <limits>
-#else
-using uint16_t      = unsigned short;
 #endif
 
 namespace common {

--- a/src/backend/common/half.hpp
+++ b/src/backend/common/half.hpp
@@ -16,6 +16,7 @@
 #include <backend.hpp>
 
 #ifndef __CUDACC_RTC__
+#include <af/compilers.h>
 #include <cstring>
 #include <ostream>
 #include <string>
@@ -26,12 +27,6 @@
 using uint16_t      = unsigned short;
 #endif
 
-#if AF_COMPILER_CXX_RELAXED_CONSTEXPR
-#define CONSTEXPR_DH constexpr __DH__
-#else
-#define CONSTEXPR_DH __DH__
-#endif
-
 namespace common {
 
 #if defined(__CUDA_ARCH__)
@@ -40,7 +35,58 @@ using native_half_t = __half;
 using native_half_t = uint16_t;
 #endif
 
-#ifndef __CUDACC_RTC__
+#ifdef __CUDACC_RTC__
+template<typename T>
+AF_CONSTEXPR __DH__ native_half_t float2half(T value) {
+    return __float2half(value);
+}
+
+AF_CONSTEXPR __DH__ inline float half2float(native_half_t value) noexcept {
+    return __half2float(value);
+}
+
+template<typename T>
+AF_CONSTEXPR __DH__ native_half_t int2half(T value) noexcept;
+
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(int value) noexcept {
+    return __int2half_rn(value);
+}
+
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(unsigned value) noexcept {
+    return __uint2half_rn(value);
+}
+
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(long long value) noexcept {
+    return __ll2half_rn(value);
+}
+
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(unsigned long long value) noexcept {
+    return __ull2half_rn(value);
+}
+
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(short value) noexcept {
+    return __short2half_rn(value);
+}
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(unsigned short value) noexcept {
+    return __ushort2half_rn(value);
+}
+
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(char value) noexcept {
+    return __ull2half_rn(value);
+}
+template<>
+AF_CONSTEXPR __DH__ native_half_t int2half(unsigned char value) noexcept {
+    return __ull2half_rn(value);
+}
+
+#else
 
 /// Convert integer to half-precision floating point.
 ///
@@ -53,7 +99,7 @@ using native_half_t = uint16_t;
 ///
 /// \return binary representation of half-precision value
 template<std::float_round_style R, bool S, typename T>
-CONSTEXPR_DH native_half_t int2half_impl(T value) noexcept {
+AF_CONSTEXPR __DH__ native_half_t int2half_impl(T value) noexcept {
     static_assert(std::is_integral<T>::value,
                   "int to half conversion only supports builtin integer types");
     if (S) value = -value;
@@ -91,17 +137,16 @@ CONSTEXPR_DH native_half_t int2half_impl(T value) noexcept {
 template<typename T, std::float_round_style R = std::round_to_nearest,
          typename std::enable_if_t<std::is_integral<T>::value &&
                                    std::is_signed<T>::value>* = nullptr>
-CONSTEXPR_DH native_half_t int2half(T value) noexcept {
-    uint16_t out;
-    out = (value < 0) ? int2half_impl<R, true, T>(value)
-                      : int2half_impl<R, false, T>(value);
+AF_CONSTEXPR __DH__ native_half_t int2half(T value) noexcept {
+    uint16_t out = (value < 0) ? int2half_impl<R, true, T>(value)
+                               : int2half_impl<R, false, T>(value);
     return out;
 }
 
 template<typename T, std::float_round_style R = std::round_to_nearest,
          typename std::enable_if_t<std::is_integral<T>::value &&
                                    std::is_unsigned<T>::value>* = nullptr>
-CONSTEXPR_DH native_half_t int2half(T value) noexcept {
+AF_CONSTEXPR __DH__ native_half_t int2half(T value) noexcept {
     return int2half_impl<R, false, T>(value);
 }
 
@@ -114,7 +159,7 @@ CONSTEXPR_DH native_half_t int2half(T value) noexcept {
 /// \param value single-precision value
 /// \return binary representation of half-precision value
 template<std::float_round_style R = std::round_indeterminate>
-CONSTEXPR_DH native_half_t float2half_impl(float value) noexcept {
+__DH__ native_half_t float2half_impl(float value) noexcept {
     uint32_t bits = 0;  // = *reinterpret_cast<uint32*>(&value);
                         // //violating strict aliasing!
     std::memcpy(&bits, &value, sizeof(float));
@@ -249,9 +294,9 @@ CONSTEXPR_DH native_half_t float2half_impl(float value) noexcept {
 ///
 /// \return binary representation of half-precision value
 template<std::float_round_style R>
-CONSTEXPR_DH native_half_t float2half_impl(double value) {
-    uint64_t bits;  // = *reinterpret_cast<uint64*>(&value);		//violating
-                    // strict aliasing!
+__DH__ native_half_t float2half_impl(double value) {
+    uint64_t bits{0};  // = *reinterpret_cast<uint64*>(&value);		//violating
+                       // strict aliasing!
     std::memcpy(&bits, &value, sizeof(double));
     uint32_t hi = bits >> 32, lo = bits & 0xFFFFFFFF;
     uint16_t hbits = (hi >> 16) & 0x8000;
@@ -267,7 +312,7 @@ CONSTEXPR_DH native_half_t float2half_impl(double value) {
             return hbits | (0x7BFF + (hbits >> 15));
         return hbits | (0x7BFF + (R != std::round_toward_zero));
     }
-    int g, s = lo != 0;
+    int g = 0, s = lo != 0;
     if (exp > 1008) {
         g = (hi >> 9) & 1;
         s |= (hi & 0x1FF) != 0;
@@ -279,7 +324,6 @@ CONSTEXPR_DH native_half_t float2half_impl(double value) {
         s |= (hi & ((1L << i) - 1)) != 0;
         hbits |= hi >> (i + 1);
     } else {
-        g = 0;
         s |= hi != 0;
     }
     if (R == std::round_to_nearest)
@@ -296,7 +340,11 @@ CONSTEXPR_DH native_half_t float2half_impl(double value) {
 }
 
 template<typename T, std::float_round_style R = std::round_to_nearest>
-CONSTEXPR_DH native_half_t float2half(T val) {
+#ifdef __CUDA_ARCH__
+AF_CONSTEXPR
+#endif
+    __DH__ native_half_t
+    float2half(T val) {
 #ifdef __CUDA_ARCH__
     return __float2half(val);
 #else
@@ -304,12 +352,12 @@ CONSTEXPR_DH native_half_t float2half(T val) {
 #endif
 }
 
-CONSTEXPR_DH inline float half2float(native_half_t value) noexcept {
+__DH__ inline float half2float(native_half_t value) noexcept {
 #ifdef __CUDA_ARCH__
     return __half2float(value);
 #else
     // return _cvtsh_ss(data.data_);
-    uint32_t mantissa_table[2048] = {
+    constexpr uint32_t mantissa_table[2048] = {
         0x00000000, 0x33800000, 0x34000000, 0x34400000, 0x34800000, 0x34A00000,
         0x34C00000, 0x34E00000, 0x35000000, 0x35100000, 0x35200000, 0x35300000,
         0x35400000, 0x35500000, 0x35600000, 0x35700000, 0x35800000, 0x35880000,
@@ -695,7 +743,7 @@ CONSTEXPR_DH inline float half2float(native_half_t value) noexcept {
 ///           value
 /// \param value The value to convert to integer
 template<std::float_round_style R, bool E, typename T>
-T half2int(native_half_t value) {
+AF_CONSTEXPR T half2int(native_half_t value) {
     static_assert(std::is_integral<T>::value,
                   "half to int conversion only supports builtin integer types");
     unsigned int e = value & 0x7FFF;
@@ -724,58 +772,6 @@ T half2int(native_half_t value) {
     return (value & 0x8000) ? -static_cast<T>(m) : static_cast<T>(m);
 }
 
-#else
-
-template<typename T>
-CONSTEXPR_DH native_half_t float2half(T value) {
-    return __float2half(value);
-}
-
-CONSTEXPR_DH inline float half2float(native_half_t value) noexcept {
-    return __half2float(value);
-}
-
-template<typename T>
-CONSTEXPR_DH native_half_t int2half(T value) noexcept;
-
-template<>
-CONSTEXPR_DH native_half_t int2half(int value) noexcept {
-    return __int2half_rn(value);
-}
-
-template<>
-CONSTEXPR_DH native_half_t int2half(unsigned value) noexcept {
-    return __uint2half_rn(value);
-}
-
-template<>
-CONSTEXPR_DH native_half_t int2half(long long value) noexcept {
-    return __ll2half_rn(value);
-}
-
-template<>
-CONSTEXPR_DH native_half_t int2half(unsigned long long value) noexcept {
-    return __ull2half_rn(value);
-}
-
-template<>
-CONSTEXPR_DH native_half_t int2half(short value) noexcept {
-    return __short2half_rn(value);
-}
-template<>
-CONSTEXPR_DH native_half_t int2half(unsigned short value) noexcept {
-    return __ushort2half_rn(value);
-}
-
-template<>
-CONSTEXPR_DH native_half_t int2half(char value) noexcept {
-    return __ull2half_rn(value);
-}
-template<>
-CONSTEXPR_DH native_half_t int2half(unsigned char value) noexcept {
-    return __ull2half_rn(value);
-}
-
 #endif  // __CUDACC_RTC__
 
 namespace internal {
@@ -783,28 +779,28 @@ namespace internal {
 struct binary_t {};
 
 /// Tag for binary construction.
-static constexpr binary_t binary;
+static constexpr binary_t binary = binary_t{};
 }  // namespace internal
 
 class half;
 
-CONSTEXPR_DH static inline bool operator==(common::half lhs,
-                                           common::half rhs) noexcept;
-CONSTEXPR_DH static inline bool operator!=(common::half lhs,
-                                           common::half rhs) noexcept;
-CONSTEXPR_DH static inline bool operator<(common::half lhs,
-                                          common::half rhs) noexcept;
-CONSTEXPR_DH static inline bool operator<(common::half lhs, float rhs) noexcept;
-CONSTEXPR_DH static inline bool isinf(half val) noexcept;
+AF_CONSTEXPR __DH__ static inline bool operator==(common::half lhs,
+                                                  common::half rhs) noexcept;
+AF_CONSTEXPR __DH__ static inline bool operator!=(common::half lhs,
+                                                  common::half rhs) noexcept;
+__DH__ static inline bool operator<(common::half lhs,
+                                    common::half rhs) noexcept;
+__DH__ static inline bool operator<(common::half lhs, float rhs) noexcept;
+AF_CONSTEXPR __DH__ static inline bool isinf(half val) noexcept;
 
 /// Classification implementation.
 /// \param arg value to classify
 /// \retval true if not a number
 /// \retval false else
-CONSTEXPR_DH static inline bool isnan(common::half val) noexcept;
+AF_CONSTEXPR __DH__ static inline bool isnan(common::half val) noexcept;
 
 class alignas(2) half {
-    native_half_t data_;
+    native_half_t data_ = 0;
 
 #if !defined(NVCC) && !defined(__CUDACC_RTC__)
     // NVCC on OSX performs a weird transformation where it removes the std::
@@ -814,48 +810,63 @@ class alignas(2) half {
 #endif
 
    public:
-    half() = default;
+    AF_CONSTEXPR half() = default;
 
     /// Constructor.
     /// \param bits binary representation to set half to
-    CONSTEXPR_DH half(internal::binary_t, uint16_t bits) noexcept : data_() {
-        memcpy(&data_, &bits, sizeof(uint16_t));
+    AF_CONSTEXPR __DH__ half(internal::binary_t, uint16_t bits) noexcept
+        :
+#if defined(__CUDA_ARCH__)
+        data_(__ushort_as_half(bits))
+#else
+        data_(bits)
+#endif
+    {
     }
 
-    CONSTEXPR_DH explicit half(double value) noexcept
+#if defined(__CUDA_ARCH__)
+    AF_CONSTEXPR
+#endif
+    __DH__ explicit half(double value) noexcept
         : data_(float2half<double>(value)) {}
 
-    CONSTEXPR_DH explicit half(float value) noexcept
+#if defined(__CUDA_ARCH__)
+    AF_CONSTEXPR
+#endif
+    __DH__ explicit half(float value) noexcept
         : data_(float2half<float>(value)) {}
 
-#ifndef __CUDA_RTC__
     template<typename T>
-    CONSTEXPR_DH explicit half(T value) noexcept : data_(int2half<T>(value)) {}
+    AF_CONSTEXPR __DH__ explicit half(T value) noexcept
+        : data_(int2half<T>(value)) {}
 
-    CONSTEXPR_DH half& operator=(const double& value) noexcept {
+#if defined(__CUDA_ARCH__)
+    AF_CONSTEXPR
+#endif
+    __DH__ half& operator=(const double& value) noexcept {
         data_ = float2half<double>(value);
         return *this;
     }
-#endif
 
 #if defined(__CUDA_ARCH__)
-    CONSTEXPR_DH explicit half(const __half& value) noexcept : data_(value) {}
-    CONSTEXPR_DH half& operator=(__half&& value) noexcept {
+    AF_CONSTEXPR __DH__ explicit half(const __half& value) noexcept
+        : data_(value) {}
+    AF_CONSTEXPR __DH__ half& operator=(__half&& value) noexcept {
         data_ = value;
         return *this;
     }
 #endif
 
-    CONSTEXPR_DH explicit operator float() const noexcept {
+    __DH__ explicit operator float() const noexcept {
         return half2float(data_);
     }
 
-    CONSTEXPR_DH explicit operator double() const noexcept {
+    __DH__ explicit operator double() const noexcept {
         // TODO(umar): convert directly to double
         return half2float(data_);
     }
 
-    CONSTEXPR_DH explicit operator short() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator short() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2short_rn(data_);
 #else
@@ -863,7 +874,7 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH explicit operator long long() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator long long() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2ll_rn(data_);
 #else
@@ -871,7 +882,7 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH explicit operator int() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator int() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2int_rn(data_);
 #else
@@ -879,7 +890,7 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH explicit operator unsigned() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator unsigned() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2uint_rn(data_);
 #else
@@ -887,7 +898,7 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH explicit operator unsigned short() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator unsigned short() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2ushort_rn(data_);
 #else
@@ -895,7 +906,7 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH explicit operator unsigned long long() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator unsigned long long() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2ull_rn(data_);
 #else
@@ -903,7 +914,7 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH explicit operator char() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator char() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2short_rn(data_);
 #else
@@ -911,7 +922,7 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH explicit operator unsigned char() const noexcept {
+    AF_CONSTEXPR __DH__ explicit operator unsigned char() const noexcept {
 #ifdef __CUDA_ARCH__
         return __half2short_rn(data_);
 #else
@@ -920,18 +931,17 @@ class alignas(2) half {
     }
 
 #if defined(__CUDA_ARCH__)
-    CONSTEXPR_DH operator __half() const noexcept { return data_; };
+    AF_CONSTEXPR __DH__ operator __half() const noexcept { return data_; };
 #endif
 
-    friend CONSTEXPR_DH bool operator==(half lhs, half rhs) noexcept;
-    friend CONSTEXPR_DH bool operator!=(half lhs, half rhs) noexcept;
-    friend CONSTEXPR_DH bool operator<(common::half lhs,
-                                       common::half rhs) noexcept;
-    friend CONSTEXPR_DH bool operator<(common::half lhs, float rhs) noexcept;
-    friend CONSTEXPR_DH bool isinf(half val) noexcept;
-    friend CONSTEXPR_DH inline bool isnan(half val) noexcept;
+    friend AF_CONSTEXPR __DH__ bool operator==(half lhs, half rhs) noexcept;
+    friend AF_CONSTEXPR __DH__ bool operator!=(half lhs, half rhs) noexcept;
+    friend __DH__ bool operator<(common::half lhs, common::half rhs) noexcept;
+    friend __DH__ bool operator<(common::half lhs, float rhs) noexcept;
+    friend AF_CONSTEXPR __DH__ bool isinf(half val) noexcept;
+    friend AF_CONSTEXPR __DH__ inline bool isnan(half val) noexcept;
 
-    CONSTEXPR_DH common::half operator-() const {
+    AF_CONSTEXPR __DH__ common::half operator-() const {
 #if __CUDA_ARCH__ >= 530
         return common::half(__hneg(data_));
 #elif defined(__CUDA_ARCH__)
@@ -941,11 +951,17 @@ class alignas(2) half {
 #endif
     }
 
-    CONSTEXPR_DH common::half operator+() const { return *this; }
+    AF_CONSTEXPR __DH__ common::half operator+() const { return *this; }
+
+    AF_CONSTEXPR static half infinity() {
+        half out;
+        out.data_ = 0x7C00;
+        return out;
+    }
 };
 
-CONSTEXPR_DH static inline bool operator==(common::half lhs,
-                                           common::half rhs) noexcept {
+AF_CONSTEXPR __DH__ static inline bool operator==(common::half lhs,
+                                                  common::half rhs) noexcept {
 #if __CUDA_ARCH__ >= 530
     return __heq(lhs.data_, rhs.data_);
 #elif defined(__CUDA_ARCH__)
@@ -956,8 +972,8 @@ CONSTEXPR_DH static inline bool operator==(common::half lhs,
 #endif
 }
 
-CONSTEXPR_DH static inline bool operator!=(common::half lhs,
-                                           common::half rhs) noexcept {
+AF_CONSTEXPR __DH__ static inline bool operator!=(common::half lhs,
+                                                  common::half rhs) noexcept {
 #if __CUDA_ARCH__ >= 530
     return __hne(lhs.data_, rhs.data_);
 #else
@@ -965,8 +981,8 @@ CONSTEXPR_DH static inline bool operator!=(common::half lhs,
 #endif
 }
 
-CONSTEXPR_DH static inline bool operator<(common::half lhs,
-                                          common::half rhs) noexcept {
+__DH__ static inline bool operator<(common::half lhs,
+                                    common::half rhs) noexcept {
 #if __CUDA_ARCH__ >= 530
     return __hlt(lhs.data_, rhs.data_);
 #elif defined(__CUDA_ARCH__)
@@ -979,8 +995,7 @@ CONSTEXPR_DH static inline bool operator<(common::half lhs,
 #endif
 }
 
-CONSTEXPR_DH static inline bool operator<(common::half lhs,
-                                          float rhs) noexcept {
+__DH__ static inline bool operator<(common::half lhs, float rhs) noexcept {
 #if defined(__CUDA_ARCH__)
     return __half2float(lhs.data_) < rhs;
 #else
@@ -1067,49 +1082,49 @@ class numeric_limits<common::half> : public numeric_limits<float> {
     static constexpr int max_exponent10 = 4;
 
     /// Smallest positive normal value.
-    static CONSTEXPR_DH common::half min() noexcept {
+    static AF_CONSTEXPR __DH__ common::half min() noexcept {
         return common::half(common::internal::binary, 0x0400);
     }
 
     /// Smallest finite value.
-    static CONSTEXPR_DH common::half lowest() noexcept {
+    static AF_CONSTEXPR __DH__ common::half lowest() noexcept {
         return common::half(common::internal::binary, 0xFBFF);
     }
 
     /// Largest finite value.
-    static CONSTEXPR_DH common::half max() noexcept {
+    static AF_CONSTEXPR __DH__ common::half max() noexcept {
         return common::half(common::internal::binary, 0x7BFF);
     }
 
     /// Difference between one and next representable value.
-    static CONSTEXPR_DH common::half epsilon() noexcept {
+    static AF_CONSTEXPR __DH__ common::half epsilon() noexcept {
         return common::half(common::internal::binary, 0x1400);
     }
 
     /// Maximum rounding error.
-    static CONSTEXPR_DH common::half round_error() noexcept {
+    static AF_CONSTEXPR __DH__ common::half round_error() noexcept {
         return common::half(
             common::internal::binary,
             (round_style == std::round_to_nearest) ? 0x3800 : 0x3C00);
     }
 
     /// Positive infinity.
-    static CONSTEXPR_DH common::half infinity() noexcept {
+    static AF_CONSTEXPR __DH__ common::half infinity() noexcept {
         return common::half(common::internal::binary, 0x7C00);
     }
 
     /// Quiet NaN.
-    static CONSTEXPR_DH common::half quiet_NaN() noexcept {
+    static AF_CONSTEXPR __DH__ common::half quiet_NaN() noexcept {
         return common::half(common::internal::binary, 0x7FFF);
     }
 
     /// Signalling NaN.
-    static CONSTEXPR_DH common::half signaling_NaN() noexcept {
+    static AF_CONSTEXPR __DH__ common::half signaling_NaN() noexcept {
         return common::half(common::internal::binary, 0x7DFF);
     }
 
     /// Smallest positive subnormal value.
-    static CONSTEXPR_DH common::half denorm_min() noexcept {
+    static AF_CONSTEXPR __DH__ common::half denorm_min() noexcept {
         return common::half(common::internal::binary, 0x0001);
     }
 };
@@ -1139,19 +1154,17 @@ struct hash<common::half>  //: unary_function<common::half,size_t>
 #endif
 
 namespace common {
-CONSTEXPR_DH
-static bool isinf(half val) noexcept {
+AF_CONSTEXPR __DH__ static bool isinf(half val) noexcept {
 #if __CUDA_ARCH__ >= 530
     return __hisinf(val.data_);
 #elif defined(__CUDA_ARCH__)
     return ::isinf(__half2float(val));
 #else
-    return val == std::numeric_limits<half>::infinity() ||
-           val == -std::numeric_limits<half>::infinity();
+    return val == half::infinity() || val == -half::infinity();
 #endif
 }
 
-CONSTEXPR_DH static inline bool isnan(half val) noexcept {
+AF_CONSTEXPR __DH__ static inline bool isnan(half val) noexcept {
 #if __CUDA_ARCH__ >= 530
     return __hisnan(val.data_);
 #elif defined(__CUDA_ARCH__)

--- a/src/backend/common/unique_handle.hpp
+++ b/src/backend/common/unique_handle.hpp
@@ -8,6 +8,8 @@
  ********************************************************/
 #pragma once
 
+#include <af/compilers.h>
+
 namespace common {
 
 /// Deletes a handle.
@@ -72,7 +74,8 @@ class unique_handle {
     constexpr operator const T &() const noexcept { return handle_; }
 
     unique_handle(const unique_handle &other) noexcept = delete;
-    unique_handle(unique_handle &&other) noexcept : handle_(other.handle_) {
+    AF_CONSTEXPR unique_handle(unique_handle &&other) noexcept
+        : handle_(other.handle_) {
         other.handle_ = 0;
     }
 

--- a/src/backend/opencl/kernel/sparse_arith.hpp
+++ b/src/backend/opencl/kernel/sparse_arith.hpp
@@ -33,7 +33,7 @@ constexpr unsigned TY      = 8;
 constexpr unsigned THREADS = TX * TY;
 
 template<af_op_t op>
-constexpr std::string getOpString() {
+AF_CONSTEXPR std::string getOpString() {
     switch (op) {
         case af_add_t: return "ADD";
         case af_sub_t: return "SUB";

--- a/src/backend/opencl/types.hpp
+++ b/src/backend/opencl/types.hpp
@@ -132,20 +132,23 @@ constexpr const char *getTypeBuildDefinition() {
     using std::begin;
     using std::end;
     using std::is_same;
-    array<bool, sizeof...(ARGS)> is_half = {is_same<ARGS, half>::value...};
-    array<bool, sizeof...(ARGS) * 2> is_double = {
-        is_same<ARGS, double>::value..., is_same<ARGS, cdouble>::value...};
+    array<bool, sizeof...(ARGS)> is_half    = {is_same<ARGS, half>::value...};
+    array<bool, sizeof...(ARGS)> is_double  = {is_same<ARGS, double>::value...};
+    array<bool, sizeof...(ARGS)> is_cdouble = {
+        is_same<ARGS, cdouble>::value...};
 
     bool half_def =
         any_of(begin(is_half), end(is_half), [](bool val) { return val; });
     bool double_def =
         any_of(begin(is_double), end(is_double), [](bool val) { return val; });
+    bool cdouble_def = any_of(begin(is_cdouble), end(is_cdouble),
+                              [](bool val) { return val; });
 
-    if (half_def && double_def) {
+    if (half_def && (double_def || cdouble_def)) {
         return " -D USE_HALF -D USE_DOUBLE";
     } else if (half_def) {
         return " -D USE_HALF";
-    } else if (double_def) {
+    } else if (double_def || cdouble_def) {
         return " -D USE_DOUBLE";
     } else {
         return "";

--- a/src/backend/opencl/types.hpp
+++ b/src/backend/opencl/types.hpp
@@ -12,6 +12,7 @@
 #include <cl2hpp.hpp>
 #include <common/kernel_type.hpp>
 #include <common/traits.hpp>
+#include <af/compilers.h>
 
 #include <algorithm>
 #include <array>
@@ -125,7 +126,7 @@ inline const char *getFullName<cdouble>() {
 }  // namespace
 
 template<typename... ARGS>
-constexpr const char *getTypeBuildDefinition() {
+AF_CONSTEXPR const char *getTypeBuildDefinition() {
     using common::half;
     using std::any_of;
     using std::array;

